src/drivers/intel/dmic/dmic_nhlt.c

```diff
@@ -0,0 +1,418 @@
+
+			val = fir_cfg_b[n]->fir_control;
+			bf1 = FIR_CONTROL_B_START_GET(val);
+			bf2 = FIR_CONTROL_B_ARRAY_START_EN_GET(val);
+			bf3 = FIR_CONTROL_B_DCCOMP_GET(val);
+			bf5 = FIR_CONTROL_B_MUTE_GET(val);
+			bf6 = FIR_CONTROL_B_STEREO_GET(val);
+			dai_dbg(dai, "dmic_set_config_nhlt(): FIR_CONTROL_B = %08x", val);
+			dai_dbg(dai, "  start=%d, array_start_en=%d, dccomp=%d", bf1, bf2, bf3);
+			dai_dbg(dai, "  mute=%d, stereo=%d", bf5, bf6);
+
+			/* Clear START, set MUTE */
+			fir_control = (val & ~FIR_CONTROL_B_START_BIT) | FIR_CONTROL_B_MUTE_BIT;
+			dai_write(dai, base[n] + FIR_CONTROL_B, fir_control);
+			dai_dbg(dai, "dmic_set_config_nhlt(): FIR_CONTROL_B = %08x", fir_control);
+
+			/* Use DC_OFFSET and GAIN as such */
+			val = fir_cfg_b[n]->dc_offset_left;
+			dai_write(dai, base[n] + DC_OFFSET_LEFT_B, val);
+			dai_dbg(dai, "dmic_set_config_nhlt(): DC_OFFSET_LEFT_B = %08x", val);
+
+			val = fir_cfg_b[n]->dc_offset_right;
+			dai_write(dai, base[n] + DC_OFFSET_RIGHT_B, val);
+			dai_dbg(dai, "dmic_set_config_nhlt(): DC_OFFSET_RIGHT_B = %08x", val);
+
+			val = fir_cfg_b[n]->out_gain_left;
+			dai_write(dai, base[n] + OUT_GAIN_LEFT_B, val);
+			dai_dbg(dai, "dmic_set_config_nhlt(): OUT_GAIN_LEFT_B = %08x", val);
+
+			val = fir_cfg_b[n]->out_gain_right;
+			dai_write(dai, base[n] + OUT_GAIN_RIGHT_B, val);
+			dai_dbg(dai, "dmic_set_config_nhlt(): OUT_GAIN_RIGHT_B = %08x", val);
+		}
+
+		/* Set up FIR coefficients RAM */
+		val = pdm_cfg[n]->reuse_fir_from_pdm;
+		if (val == 0) {
+			fir_a[n] = (struct nhlt_pdm_fir_coeffs *)p;
+			p += sizeof(int32_t) * fir_length_a;
+			fir_b[n] = (struct nhlt_pdm_fir_coeffs *)p;
+			p += sizeof(int32_t) * fir_length_b;
+		} else {
+			val--;
+			if (val >= n) {
+				dai_err(dai, "dmic_set_config_nhlt(): Illegal FIR reuse 0x%x", val);
+				return -EINVAL;
+			}
+
+			if (!fir_a[val]) {
+				dai_err(dai, "dmic_set_config_nhlt(): PDM%d FIR reuse from %d fail",
+					n, val);
+				return -EINVAL;
+			}
+
+			fir_a[n] = fir_a[val];
+			fir_b[n] = fir_b[val];
+		}
+
+		if (dai->index == 0) {
+			dai_info(dai, "dmic_set_config_nhlt(): clkdiv = %d, mcic = %d, mfir_a = %d, length = %d",
+				 p_clkdiv, p_mcic, p_mfira, fir_length_a);
+			for (i = 0; i < fir_length_a; i++)
+				dai_write(dai, coef_base_a[n] + (i << 2), fir_a[n]->fir_coeffs[i]);
+		} else {
+			dai_info(dai, "dmic_set_config_nhlt(): clkdiv = %d, mcic = %d, mfir_b = %d, length = %d",
+				 p_clkdiv, p_mcic, p_mfirb, fir_length_b);
+			for (i = 0; i < fir_length_b; i++)
+				dai_write(dai, coef_base_b[n] + (i << 2), fir_b[n]->fir_coeffs[i]);
+		}
+	}
+
+	if (dai->index == 0)
+		ret = nhlt_dmic_dai_params_get(dai, out_control, pdm_cfg, fir_cfg_a);
+	else
+		ret = nhlt_dmic_dai_params_get(dai, out_control, pdm_cfg, fir_cfg_b);
+
+	if (ret)
+		return ret;
+
+	if (dai->index == 0)
+		dmic->dai_rate = DMIC_HW_IOCLK / (p_clkdiv * p_mcic * p_mfira);
+	else
+		dmic->dai_rate = DMIC_HW_IOCLK / (p_clkdiv * p_mcic * p_mfirb);
```

Comment:
ok, I commented to the previous version - it makes one a bit nervous to see 3 variables initialised to 0 and then used a dividers... :-) I verified, that they will for sure be re-assigned, and *probably* to non-0 values, e.g.
```
	clkdiv = MIC_CONTROL_PDM_CLKDIV_GET(pdm_cfg[n]->mic_control);
	p_clkdiv = clkdiv + 2;
```
and we're sure, that the loop is run at least once, so that kind of indicates, that `pclkdiv >= 2` but at least in theory without checking further `MIC_CONTROL_PDM_CLKDIV_GET()` could return `-2`. Similar for other dividers. So, I think it would be good to clarify this here. Maybe make it explicit as
```
if (dai->index == 0) {
	int div = p_clkdiv * p_mcic * p_mfira;
	if (!div) {
		dai_err(...);
		return -EINVAL;
	}
	dmic->dai_rate = DMIC_HW_IOCLK / div;
} else {
	int div = p_clkdiv * p_mcic * p_mfirb;
	if (!div) {
		dai_err(...);
		return -EINVAL;
	}
	dmic->dai_rate = DMIC_HW_IOCLK / div;
}
```
or something similar

Reply:
I agree that this does not hurt to do. A corrupt blob should not break things.